llvm/test/Transforms/InstCombine/gep-fold-chained-const-select.ll

```diff
@@ -0,0 +1,176 @@
+; CHECK-LABEL: @src_fail_different_type(
+; CHECK-NEXT:    [[V0:%.*]] = getelementptr inbounds nuw i8, ptr [[ARG1:%.*]], i64 8148
+; CHECK-NEXT:    [[V1:%.*]] = icmp sgt i32 [[ARG0:%.*]], 3
+; CHECK-NEXT:    [[V2:%.*]] = select i1 [[V1]], i64 55104, i64 21304
+; CHECK-NEXT:    [[V3:%.*]] = getelementptr i16, ptr [[V0]], i64 [[V2]]
+; CHECK-NEXT:    ret ptr [[V3]]
+;
+  %v0 = getelementptr inbounds nuw i8, ptr %arg1, i64 8148
+  %v1 = icmp sgt i32 %arg0, 3
+  %v2 = select i1 %v1, i64 55104, i64 21304
+  %v3 = getelementptr i16, ptr %v0, i64 %v2
+  ret ptr %v3
+}
+
+; Fail 2: No constant idx
+define ptr @src_fail_no_constant_idx(i32 %arg0, ptr %arg1, i64 %arg2) {
+; CHECK-LABEL: @src_fail_no_constant_idx(
+; CHECK-NEXT:    [[V0:%.*]] = getelementptr inbounds nuw i8, ptr [[ARG1:%.*]], i64 [[ARG2:%.*]]
+; CHECK-NEXT:    [[V1:%.*]] = icmp sgt i32 [[ARG0:%.*]], 3
+; CHECK-NEXT:    [[V2:%.*]] = select i1 [[V1]], i64 55104, i64 21304
+; CHECK-NEXT:    [[V3:%.*]] = getelementptr i8, ptr [[V0]], i64 [[V2]]
+; CHECK-NEXT:    ret ptr [[V3]]
+;
+  %v0 = getelementptr inbounds nuw i8, ptr %arg1, i64 %arg2
+  %v1 = icmp sgt i32 %arg0, 3
+  %v2 = select i1 %v1, i64 55104, i64 21304
+  %v3 = getelementptr i8, ptr %v0, i64 %v2
+  ret ptr %v3
+}
+
+; Fail 3: Multiple use of select
+define ptr @src_fail_select_multiple_use(i32 %arg0, ptr %arg1, ptr %arg2) {
+; CHECK-LABEL: @src_fail_select_multiple_use(
+; CHECK-NEXT:    [[V0:%.*]] = getelementptr inbounds nuw i8, ptr [[ARG1:%.*]], i64 8148
+; CHECK-NEXT:    [[V1:%.*]] = icmp sgt i32 [[ARG0:%.*]], 3
+; CHECK-NEXT:    [[V2:%.*]] = select i1 [[V1]], i64 55104, i64 21304
+; CHECK-NEXT:    [[V3:%.*]] = getelementptr i8, ptr [[V0]], i64 [[V2]]
+; CHECK-NEXT:    [[V4:%.*]] = getelementptr i8, ptr [[ARG2:%.*]], i64 [[V2]]
+; CHECK-NEXT:    store ptr [[V3]], ptr [[V4]], align 8
+; CHECK-NEXT:    ret ptr [[V3]]
+;
+  %v0 = getelementptr inbounds nuw i8, ptr %arg1, i64 8148
+  %v1 = icmp sgt i32 %arg0, 3
+  %v2 = select i1 %v1, i64 55104, i64 21304
+  %v3 = getelementptr i8, ptr %v0, i64 %v2
+  %v4 = getelementptr i8, ptr %arg2, i64 %v2
+  store ptr %v3, ptr %v4, align 8
+  ret ptr %v3
+}
+
+; Fail 4: Multiple use of source GEP
+define ptr @src_fail_source_gep_multiple_use(i32 %arg0, ptr %arg1, ptr %arg2) {
+; CHECK-LABEL: @src_fail_source_gep_multiple_use(
+; CHECK-NEXT:    [[V1:%.*]] = icmp sgt i32 [[ARG0:%.*]], 3
+; CHECK-NEXT:    [[TMP1:%.*]] = select i1 [[V1]], i64 63252, i64 29452
+; CHECK-NEXT:    [[V3:%.*]] = getelementptr i8, ptr [[ARG1:%.*]], i64 [[TMP1]]
+; CHECK-NEXT:    [[V4:%.*]] = getelementptr i8, ptr [[ARG1]], i64 8248
+; CHECK-NEXT:    store ptr [[V3]], ptr [[V4]], align 8
+; CHECK-NEXT:    ret ptr [[V3]]
+;
+  %v0 = getelementptr inbounds nuw i8, ptr %arg1, i64 8148
+  %v1 = icmp sgt i32 %arg0, 3
+  %v2 = select i1 %v1, i64 55104, i64 21304
+  %v3 = getelementptr i8, ptr %v0, i64 %v2
+  %v4 = getelementptr i8, ptr %v0, i64 100
+  store ptr %v3, ptr %v4, align 8
+  ret ptr %v3
+}
```

Comment:
I don't think this shows a proper multi-use pattern, as the two constant offset GEPs can be combined. Can you please use `call void @use(ptr %v0)` for the extra use?

Reply:
Changed.